osu.Game/Screens/Edit/Changes/IRevertibleChange.cs

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Game.Screens.Edit.Changes
+{
+    /// <summary>
+    /// Represents a change which can be undone.
+    /// </summary>
+    public interface IRevertibleChange
+    {
+        /// <summary>
+        /// Applies this change to the current state.
+        /// </summary>
+        void Apply();
+
+        /// <summary>
+        /// Applies the inverse of this change to the current state.
+        /// </summary>
+        void Revert();
+    }
+
+    public static class IRevertibleChangeExtension
+    {
+        public static void Submit(this IRevertibleChange change, NewBeatmapEditorChangeHandler? changeHandler, bool commitImmediately = false)
+        {
+            if (changeHandler != null)
+                changeHandler.Submit(change, commitImmediately);
+            else
+                change.Apply();
+        }
```

Comment:
This is a bit weird to me in general. Why do changes submit themselves to the handler like this? Reads backwards. `changeHandler.Submit(command)` reads much more natural to me.

Reply:
I eventually went with this because of what's pointed out in https://github.com/ppy/osu/pull/30314#discussion_r1803720644 its the least likely to shoot yourself in the foot.

We have a weird situation because commands/changes have to `Apply` to do anything or `Submit` to a change handler which in turn calls `Apply`.
The change handlers are often nullable so we want to `Submit` if it's not null and `Apply` if it's null.

I previously had an extension method like below that handles this logic and follows the reading direction you say feels more natural.
```cs
        public static void SafeSubmit(this NewBeatmapEditorChangeHandler? manager, IRevertableChange command, bool commitImmediately = false)
        {
            if (manager != null)
                manager.Submit(command, commitImmediately);
            else
                command.Apply();
        }
```
The issue is that you can accidentally add in a `?.` operator to the nullable change handler which would cause it to not run anything if the change handler is null.
```cs
// Wrong, does not apply the command when null
changeHandler?.SafeSubmit(...);
```
I also personally prefer writing code with the `new Command(...).Submit(changeHandler)` syntax because you can focus on the important bits first which is defining the command and then write the boilerplate, and its less nesting of parentheses.